stubs/python-jose/jose/jws.pyi

```diff
@@ -1,7 +1,9 @@
 from typing import Any
 
-def sign(payload, key, headers: Any | None = ..., algorithm=...): ...
-def verify(token, key, algorithms, verify: bool = ...): ...
-def get_unverified_header(token): ...
-def get_unverified_headers(token): ...
-def get_unverified_claims(token): ...
+def sign(
+    payload: str | dict[str, Any], key: str | dict[str, Any], headers: dict[str, Any] | None = ..., algorithm: str = ...
```

Comment:
Argh I just did a quick check on the docs, should have been a bit more careful about the implementation details, apologies! See comments below and fixes here: https://github.com/python/typeshed/pull/8238/commits/a7508a80db5fb848ea59189280ee8c7fe4f63036
- True, fixed!
- The check for `key` is actually for is _not_ an instance of `Key`, so then it calls `jwk.construct()`, which does check for a `dict` before creating a `Key` instance: https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwk.py#L70 So I kept it as `str | dict[str, Any] | Key`. Happy to drop `Key` though, as it feels a bit too implementation specific, also considering their docs only mention str or dict, see here: https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L17 (same for `jwt` changes), wdyt?
- True, fixed!

Besides, I also realised that in `jwt.decode` `algorithms` is defined as `str | Container[str] | None` https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwt.py#L56, even though it is later used in `jws.verify`, which does not accept `None`. What is normally the process for these cases? Should we "fix" the typing by not accepting `None`, or do we just follow the implementation, which does indeed have a `None` default? (even though their docs actually say `str | list` only, see https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwt.py#L63).

Reply:
> Argh I just did a quick check on the docs, should have been a bit more careful about the implementation details, apologies!

Docs can be deceiving, especially if the codebase is entirely untyped! 😄

> * The check for `key` is actually for is _not_ an instance of `Key`, so then it calls `jwk.construct()`, which does check for a `dict` before creating a `Key` instance: https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwk.py#L70 So I kept it as `str | dict[str, Any] | Key`.

Good point! Having `dict` for an argument type annotation is usually a code smell, but in this case it seems it is the correct hint. Can you add a comment to explain why we _do_ have to use `dict` in this situation rather than `Mapping`?

> Besides, I also realised that in `jwt.decode` `algorithms` is defined as `str | Container[str] | None` https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jwt.py#L56, even though it is later used in `jws.verify`, which does not accept `None`.

Hmm, it looks like it _is_ actually safe to pass `None` to the `algorithms` parameter of `jws.verify()`, since `jws.verify()` passes it to `_verify_signature`, and `_verify_signature` then does an `if algorithms is not None` check here: https://github.com/mpdavis/python-jose/blob/96474ecfb6ad3ce16f41b0814ab5126d58725e2a/jose/jws.py#L256-L257. So I think we should change the `algorithms` parameter of `jws.verify()` to be `str | Container[str] | None`, and add a comment about why `None` is in the union.

Reply:
Sounds good, done! Let me know if you'd like to rephrase anything 🙂  https://github.com/python/typeshed/pull/8238/commits/48c8e75dcd744aa48d9dfec7794ea74569af8d3a